src/platform/x86/x86_64/gdt.c

```diff
@@ -0,0 +1,178 @@
+	/* 
+	 * Be careful to have the correct tss structure in memory and in gdt,
+	 * qemu does not check tss validity, but real machine does check 
+	 */
+	asm volatile("ltr %w0" : : "r"(SEL_TSS));
+	flash_selectors();
+}
+
+/* System segment or code/data segment? */
+enum seg_class
+{
+	CLS_SYSTEM    = 0, /* System segment. */
+	CLS_CODE_DATA = 1  /* Code or data segment. */
+};
+
+/* Limit has byte or 4 kB page granularity? */
+enum seg_granularity
+{
+	GRAN_BYTE = 0, /* Limit has 1-byte granularity. */
+	GRAN_PAGE = 1  /* Limit has 4 kB granularity. */
+};
+
+
+static void
+make_gdtr_operand(u8_t gdtr_addr[10], u16_t limit, u64_t base)
+{
+	*((u16_t*)(&gdtr_addr[0])) = limit;
+	*((u64_t*)(&gdtr_addr[2])) = base;
+}
+
+static u64_t
+make_code_desc(int dpl)
+{
+	/* 
+	 * make sure to init variables before use, you don't know 
+	 * whether the value comes from registers or memory, if it does 
+	 * comes from memory, value in that address could be random, 
+	 * thus you need to init that memory first. 	
+	 */
+	u32_t e0 = 0, e1 = 0;
+
+	e1 = ((0)
+	      | (10 << 8)       /* Segment type. */
+	      | (1 << 12)       /* 0=system, 1=code/data. */
+	      | (dpl << 13)     /* Descriptor privilege. */
+	      | (1 << 15)       /* Present. */
+		  | (1 << 21)       /* L bit. */
+	      | (0 << 22));       /* D/B bit. */
+
+	return e0 | ((u64_t)e1 << 32);
+}
+
+
+static u64_t
+make_data_desc(int dpl)
+{
+	/* 
+	 * make sure to init variables before use, you don't know 
+	 * whether the value comes from registers or memory, if it does 
+	 * comes from memory, value in that address could be random, 
+	 * thus you need to init that memory first. 	
+	 */
+	u32_t e0 = 0, e1 = 0;
+
+	e1 = ((0)
+	      | (2 << 8)        /* Segment type. */
+	      | (1 << 12)       /* 0=system, 1=code/data. */
+	      | (dpl << 13)     /* Descriptor privilege. */
+	      | (1 << 15));     /* Present. */
+
+	return e0 | ((u64_t)e1 << 32);
+}
+
+static void
+make_tss_desc(volatile u64_t * tss_desc, u64_t tss_addr)
+{
+	/* 
+	 * make sure to init variables before use, you don't know 
+	 * whether the value comes from registers or memory, if it does 
+	 * comes from memory, value in that address could be random, 
+	 * thus you need to init that memory first. 	
+	 */
+	u32_t e0 = 0, e1 = 0;
+	u64_t e2 = (tss_addr >> 32) & 0x00000000ffffffff;
+	u32_t base = (u32_t)tss_addr;
+	int type = 9;
+	u32_t limit = 0x67;
+	e0 = (( limit & 0xffff) 		/* Limit 15:0. */
+	      | (base << 16));			/* Base 15:0. */
+
+	e1 = (((base >> 16) & 0xff)		/* Base 23:16. */
+	      | (type << 8)				/* Segment type. */
+	      | (0 << 12)				/* 0=system, 1=code/data. */
+	      | (0 << 13)				/* Descriptor privilege. */
+	      | (1 << 15)             	/* Present. */
+	      | (limit & 0xf0000)     	/* Limit 16:19. */
+	      | (base & 0xff000000)); 	/* Base 31:24. */
+
+	*tss_desc = (e0 | ((u64_t)e1 << 32));	
+	*(tss_desc + 1) = e2;
+}
+
+static void
+flash_selectors(void)
```

Comment:
I don't know what this does from the name. The assembly isn't very self-documenting, so a good name, and potentially a comment is pretty necessary. What problem is this solving that isn't necessary on x86-32 (or is this adapted from there?).